File: src/main/Java/dfs/boj1707.java
import java.io.*;
import java.util.ArrayList;
import java.util.StringTokenizer;
import java.util.stream.IntStream;

public class boj1707 {
    private static ArrayList<Integer>[] linked;
    private static boolean[] isVisited;
    private static int[] check;
    private static boolean isBipartite;

    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        int k = Integer.parseInt(reader.readLine());
        while (k-- > 0) {
            isBipartite = true;
            StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
            int v = Integer.parseInt(tokenizer.nextToken());
            int e = Integer.parseInt(tokenizer.nextToken());
            linked = new ArrayList[v + 1];
            IntStream.range(1, v + 1).forEach(x -> linked[x] = new ArrayList<>());
            isVisited = new boolean[v + 1];
            check = new int[v + 1];
            while (e-- > 0) {
                tokenizer = new StringTokenizer(reader.readLine());
                int x = Integer.parseInt(tokenizer.nextToken());
                int y = Integer.parseInt(tokenizer.nextToken());
                linked[x].add(y);
                linked[y].add(x);
            }
            for (int i = 1; i <= v; i++) {
                if (isBipartite) dfs(i);
                else break;
            }
            writer.append((isBipartite) ? "YES" : "NO").append('\n');
        }
        writer.close();
    }

    private static void dfs(int n) {
        isVisited[n] = true;
        linked[n].forEach(x -> {
            if (!isVisited[x]) {
                check[x] = (check[n] + 1) % 2;
                dfs(x);
            } else if (check[n] == check[x]) isBipartite = false;
        });
    }
}

File: src/main/kotlin/dfs/boj1707.kt

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.stream.IntStream

private var linked = arrayOf(ArrayList<Int>())
private var isVisited = booleanArrayOf()
private var check = intArrayOf()
private var isBipartite = true
fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    val k = r.readLine().toInt()
    repeat(k) {
        isBipartite = true
        val (v, e) = r.readLine().split(" ").map { it.toInt() }
        linked = Array(v + 1) { ArrayList() }
        isVisited = BooleanArray(v + 1)
        check = IntArray(v + 1)
        repeat(e) {
            val (x, y) = r.readLine().split(" ").map { it.toInt() }
            linked[x].add(y)
            linked[y].add(x)
        }
        IntStream.range(1, v + 1).forEach {
            if (isBipartite) dfs(it)
            else return@forEach
        }
        w.appendLine(if (isBipartite) "YES" else "NO")
    }
    w.close()
}

private fun dfs(n: Int) {
    isVisited[n] = true
    linked[n].forEach {
        if (!isVisited[it]) {
            check[it] = (check[n] + 1) % 2
            dfs(it)
        } else if (check[n] == check[it]) isBipartite = false
    }
}